File: Day_6/aoc_day6.cpp
// See: Advent of Code 2021, Day 6: Lanternfish
// Made by: krauluk1
#include <iostream>
#include <fstream>
#include <string>
#include <cstring>
#include <sstream>
#include <vector>


class Laternfish
{
    private:
        std::vector<int> group;
        std::vector<int> process_string(const std::string str);

    public:
        Laternfish(std::string path);
        std::vector<int> read_txt(const std::string path);
        long long count_fishes(int days);
};

Laternfish::Laternfish(std::string path){
    group = read_txt(path);
}

std::vector<int> Laternfish::read_txt(const std::string path){
    std::ifstream infile(path);

    std::string line;
    std::vector<int> vector_full;
    std::vector<int> vector_line;

    while (infile >> line)
    {
        vector_line = process_string(line);
        vector_full.insert(vector_full.end(), vector_line.begin(), vector_line.end());
    }
    return vector_full;
}

std::vector<int> Laternfish::process_string(std::string st){
    std::stringstream full_str(st);
    std::string segment;
    std::vector<int> seglist;

    while(std::getline(full_str, segment, ','))
    {
        seglist.push_back(std::stoi(segment));
    }

    return seglist;
}

long long Laternfish::count_fishes(int days){
    // Initialize Counter
    // Need a big datatype, size of int not enough
    std::vector<long long> number(9, 0);

    for(int i:group){
        number[i]+=1;
    }

    for(int i = 1; i < days+1; i++){
        long long zeros = number[0];
        number.erase(number.begin());
        number.push_back(zeros);
        number[6] += zeros;
    }  

    long long sum = 0;
    for(long long i:number){
        sum+=i;
    }
    return sum;
}


int main()
{
    Laternfish d("input.txt");
    std::cout << "Solution Task 1: " << d.count_fishes(80) << std::endl;
    std::cout << "Solution Task 2: " << d.count_fishes(256) << std::endl;
    return 0;
}

File: Day_6/aoc_day6.py
# See: Advent of Code 2021, Day 6: Lanternfish
# Made by: krauluk1
import numpy

class Lanternfish(object):
    def __init__(self, path):
        # Prepare Data
        self.data_start = self.__read_txt(path)

    def __read_txt(self, path):
        # Get TXT as list
        with open(path, "r") as f:
            line = f.read().splitlines()
            for i in range(len(line)):
                line[i] = list(map(int,line[i].split(',')))
            return numpy.array(line).ravel().tolist()

    # -------------
    # much faster and more effective
    def calculate_fish_nr2(self, days):
        # 0 till max 8 days
        counter =[0]*9
        data_stop = self.data_start[:]

        # Fill counter
        for i in data_stop:
            counter[i]+=1

        for day in range(1, days+1):
            zeros = counter.pop(0)
            # New fishes
            counter.append(zeros)
            # Reset counter for normal fishes
            counter[6] += zeros

        return numpy.sum(counter)

    
    # -------------
    # not fast and effective, but the same output as in the example.
    def __new_day(self, data):
        for fish in range(len(data)-1, -1, -1):
            if(data[fish] == 0):
                # Reset timer
                data[fish] = 7
                # Create new fish
                data = numpy.append(data, 8)
            data[fish] -= 1
        return data

    def calculate_fish_nr1(self, days):
        data_stop = self.data_start[:]
        print("Initial state:", data_stop)

        for i in range(1, days+1):
            data_stop = self.__new_day(data_stop)
            print("After", i, "day:", data_stop)
        return len(data_stop)
    # ------------------

h = Lanternfish('Day_6\input.txt')
print("Answer taks 1:", h.calculate_fish_nr2(80))
print("Answer taks 2:", h.calculate_fish_nr2(256))